### ISSUE TITLE
Cache dynamically added rules with UrlManager::addRules()

### ISSUE THREAD

**Issue by yuriystovbur** [yiisoft/yii2#13425]
Compiled URL rules will be cached only on UrlManager::init() stage.
Additional compiled URL rules are not cached.

**Comment by SilverFire** [comment]
Could you explain your use case?

**Comment by scriptcube** [comment]
@SilverFire, URL rules can be dynamically added to the [[yii\web\UrlManager|URL manager]] https://github.com/yiisoft/yii2/blob/master/docs/guide/runtime-routing.md#user-content-adding-rules-dynamically-

But URL rules added by UrlManager::addRules() will be compiled and not will be cached.

```php
    public function addRules($rules, $append = true)
    {
        if (!$this->enablePrettyUrl) {
            return;
        }
        $rules = $this->buildRules($rules);
        if ($append) {
            $this->rules = array_merge($this->rules, $rules);
        } else {
            $this->rules = array_merge($rules, $this->rules);
        }
    }
```

Compiled URL rules will be cached only on UrlManager::init() stage:

```php
    /**
     * Initializes UrlManager.
     */
    public function init()
    {
        //...
        if ($this->cache instanceof Cache) {
            $cacheKey = $this->cacheKey;
            $hash = md5(json_encode($this->rules));
            if (($data = $this->cache->get($cacheKey)) !== false && isset($data[1]) && $data[1] === $hash) {
                $this->rules = $data[0];
            } else {
                $this->rules = $this->buildRules($this->rules);
                $this->cache->set($cacheKey, [$this->rules, $hash]);
            }
        } else {
            $this->rules = $this->buildRules($this->rules);
        }
    }
```

**Comment by SilverFire** [comment]
Rules caching requires building, serialization and setting to the cache:

```php
$hash = md5(json_encode($this->rules));
$this->rules = $this->buildRules($this->rules);
$this->cache->set($cacheKey, [$this->rules, $hash]);
```

In case you add your rules dynamically, this operation will be performed for each rule. Of course it will happen only once and then you will have all intermediate steps stored in your cache, but I doubt it's a right way.

Why can't you pre-configure the URL manager instead of adding rules dynamically?

**Comment by scriptcube** [comment]
I and others developers use a modular application where each module adds its URL rules.
If many loadable modules in the application, use of cache for URL Rules is more productive.

**Comment by SilverFire** [comment]
> each module adds its URL rules

It could be done with configuration merging instead of bootstrapping.

However, the use case may be marked as valid. Thank you for your explanation.

**Comment by TerraSkye** [comment]
i have another usecase,

My application is plugin based. plugins can be seen as modules. and a plugin service registers all the activated plugins. ( dependend per user ) 
therefor the registerd urls are different per user.

**Comment by scriptcube** [comment]
https://github.com/yiisoft/yii2/blob/master/docs/guide/structure-modules.md#best-practices-
**Best Practices.** "Modules are best used in large applications whose features can be divided into several groups, each consisting of a set of closely related features."

Each module can be a set of its URL rules. Or third-party modules included to application.

I think necessary caching additional URL rules

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/compare/master...yuriystovbur:13425-fix-url-rules-cache

**Comment by SilverFire** [comment]
Fixed in https://github.com/yiisoft/yii2/commit/a1b5f077e08939382719ef70c29f9c2b78ecc987

**Comment by hightman** [comment]
@SilverFire This implementation has serious drawbacks, when added rules contain objects extended from`\yii\web\CompositeUrlRule`, they may have properties related to big object, such as Module.

These objects may not be serialized, or take up a lot of space.

**Comment by SilverFire** [comment]
You can configure properties serialization and unserialization with __sleep and __wakeup magic methods

**Comment by hightman** [comment]
In fact, I think it is not meaningful for caching added rule objects, those rules are generally generated by calling `createRules()` to dynamically.

**Comment by SilverFire** [comment]
There opposite opinion was a starting point of this issue :)

**Comment by hightman** [comment]
In addition, the `$ruleDeclarations` may contain other external objects defined by third-party or core framework, they are all serialized as caching key.

Or add a third parameter to `UrlManager::addRules` as caching option. 

My current project encountered this problem, I have to turn off caching of whole UrlManager.

**Comment by SilverFire** [comment]
Ok, could you open a new issue where we'll discuss `cacheDynamicallyAddedRules` property?
